File: 2014/04/29-MCMC.R
# Code modified from Dave Gile's post:
# http://davegiles.blogspot.ca/2014/03/mcmc-for-econometrics-students-ii_18.html

gibs <- function(seed=123,
                 nrep=105000,
                 nb  =5000,
                 rho =.5){
  
  set.seed(seed)
  y1 <- rnorm(nrep,0,sd)
  y2 <- rnorm(nrep,0,sd)
  
  sd<- sqrt(1-rho^2)
  
  # Start the Gibbs sampler loop
  for (i in 2:nrep) {
    y2 <- y2[i]+rho*y1[i-1]
    y1 <- y1[i]+rho*y2[i]
  }
  return(list(y1=yy1[(nb+1):nrep],y2=yy2[(nb+1):nrep]))
}

gibs2 <- function(seed=123,
                  nrep=105000,
                  nb  =5000,
                  rho =.5){
  
  set.seed(seed)
  
  sd<- sqrt(1-rho^2)
  
  y1 <- rnorm(nrep,0,sd)
  y2 <- rnorm(nrep,0,sd)
  
  # Start the Gibbs sampler loop
  for (i in 2:nrep) {
    y2 <- y2[i]+rho*y1[i-1]
    y1 <- y1[i]+rho*y2[i]
  }
  return(list(y1=yy1[(nb+1):nrep],y2=yy2[(nb+1):nrep]))
}
system.time(yR <- gibs())
#   user  system elapsed 
#   0.97    0.06    1.17

system.time(yR <- gibs2())
#   user  system elapsed 
#   0.35    0.00    0.36 

system.time(yR <- gibs(nrep=10^6))
#   user  system elapsed 
#   9.31    0.02    9.64 

system.time(yR <- gibs2(nrep=10^6))
#   user  system elapsed 
#   3.66    0.01    3.80 

File: 2014/04/29-MCMC.jl
# Set random seed for the random generator
#  so that our results are reproducible

function gibs(nrep=105000;
              seed=123,
              nb  =5000,
              ρ   =.5)
   # Seed - sets the random number generator seed
   #   so that it is easy to create reproducible results.
   # nrep - in the number of repetitions of the MCMC draws.
   # nb   - number of observations for the "Burn-in"
   # ρ    - Correlation between Y1 and Y2
   srand(seed)

   # Create empty vectors of length nrep
   yy₁ = yy₂ = spzeros(nrep,1)

   # Calculate standard deviation of Y2
   σ   = sqrt(1-ρ^2)

   # Initialize y1 with a single normal draw
   y₁ = randn()*σ

   # Now we start the Gibbs sampler loop
   for i in 1:nrep
     y₂ = randn()*σ + ρ*y₁
     y₁ = randn()*σ + ρ*y₂
     yy₁[i]=y₁
     yy₂[i]=y₂
   end

   return (yy₁[(nb+1):end,1],yy₂[(nb+1):end,1])
end
gibs()

# Set random seed for the random generator
#  so that our results are reproducible
function gibs2(nrep=105000;
               seed=123,
               nb  =5000,
               ρ   =.5)
   # Seed - sets the random number generator seed
   #   so that it is easy to create reproducible results.
   # nrep - in the number of repetitions of the MCMC draws.
   # nb   - number of observations for the "Burn-in"
   # ρ    - Correlation between Y1 and Y2
   srand(seed)

   # Calculate standard deviation of Y2
   sigma = sqrt(1-ρ^2)

   # Create empty vectors of length nrep
   y1 = randn(nrep)
   y2 = randn(nrep)

   # Now we start the Gibbs sampler loop
   for i in 2:nrep
     y2[i] += y1[i-1]*ρ
     y1[i] += y2[i]*ρ
   end

   return (y1[(nb+1):end,1],y2[(nb+1):end,1])
end
gibs2()

@elapsed yy = gibs()
# 0.063151467
@elapsed gibs(10^6)
# 0.479542057

@elapsed yy = gibs2()
# 0.010729382
@elapsed yy = gibs2(10^6)
# 0.065821774
